### PR TITLE
jsonrpc: Implement debug_getBadBlocks

### PR DESCRIPTION
**Pull request by sudeepdino008**
https://github.com/erigontech/erigon/issues/8171

- reference of rpc api query params/return value is taken from geth [here](https://github.com/ethereum/go-ethereum/blob/b4d99e39176f89bb0ffbb32635e9cf17b5fd7367/eth/api_debug.go#L107)
- difference is that geth returns ALL bad blocks, this PR returns 100.

### CODE REVIEW

File: erigon-lib/kv/tables.go [hunk 1]
```diff
@@ -239,8 +239,8 @@ const (
 	DatabaseInfo = "DbInfo"
 
 	// Naming:
-	//   NeaderNumber - Ethereum-specific block number. All nodes have same BlockNum.
-	//   NeaderID - auto-increment ID. Depends on order in which node see headers.
+	//   HeaderNumber - Ethereum-specific block number. All nodes have same BlockNum.
+	//   HeaderID - auto-increment ID. Depends on order in which node see headers.
```
Comment (Ethiqque) [inline erigon-lib/kv/tables.go:243]: ci falls

**Comment by somnathb1** [comment]
`getbadBlocks` does get hit quite often by some CLs, about once every block or so, i think.
It does make sense to limit output to 100 blocks in the LRU, but reading the whole table on every call doesn't make sense. There should be an in-memory cache. Assuming there are about 1000 bad blocks, it's a 1000 sequential mdbx reads - that is not bad as such, but only if it's called every block
In `accessors_chain` tho, i don't see a natural point to include this LRU cache 
It may help to add another table with `blockNum + hash` -> `{}`

**Comment by sudeepdino008** [comment]
> `getbadBlocks` does get hit quite often by some CLs, about once every block or so, i think. It does make sense to limit output to 100 blocks in the LRU, but reading the whole table on every call doesn't make sense. There should be an in-memory cache. Assuming there are about 1000 bad blocks, it's a 1000 sequential mdbx reads - that is not bad as such, but only if it's called every block In `accessors_chain` tho, i don't see a natural point to include this LRU cache It may help to add another table with `blockNum + hash` -> `{}`

that's interesting. I thought it was just one-off queries by the user. Is there a reference which I can see about CL calling getBadBlocks?  
If last 100 blocks is good enough, I think LRU cache is enough optimization. Adding new table is useful only if there are large number of bad blocks, and the assumption is that there aren't.

**Comment by somnathb1** [comment]
> > `getbadBlocks` does get hit quite often by some CLs, about once every block or so, i think. It does make sense to limit output to 100 blocks in the LRU, but reading the whole table on every call doesn't make sense. There should be an in-memory cache. Assuming there are about 1000 bad blocks, it's a 1000 sequential mdbx reads - that is not bad as such, but only if it's called every block In `accessors_chain` tho, i don't see a natural point to include this LRU cache It may help to add another table with `blockNum + hash` -> `{}`
> 
> that's interesting. I thought it was just one-off queries by the user. Is there a reference which I can see about CL calling getBadBlocks? If last 100 blocks is good enough, I think LRU cache is enough optimization. Adding new table is useful only if there are large number of bad blocks, and the assumption is that there aren't.

I think you will see those hits by running against a CL like lighthouse, nimbus, or teku against any network. You can also run a kurtosis env. Some of the CLs might also be optimizing on their side how often they call for this json resp, but in the worst case it could be bad.

**Review by yperbasis — APPROVED** [review]

File: core/rawdb/accessors_chain.go [hunk 2]
```diff
@@ -87,6 +89,25 @@ func TruncateCanonicalHash(tx kv.RwTx, blockFrom uint64, markChainAsBad bool) er
 	return nil
 }
 
+func GetLatestBadBlocks(tx kv.Tx, limit int) ([]*types.Block, error) {
+	// BadHeaderNumber table is typically small, so that we can go through all of it.
+	bheap := utils.NewBlockMaxHeap(limit)
```
Comment (AskAlexSharov) [inline core/rawdb/accessors_chain.go:94]: Probably you don't need `heap`:
- table sorted by key. key starts from blockNum.
- can read table backward - 2 ways: `Cursor` has `.Last()/.Prev()` methods. Also `tx` has `.Range` method with `order.Desc` parameter. Both ways allow read `N` keys or until some prefix.

Comment (sudeepdino008) [reply core/rawdb/accessors_chain.go:94]: the key is hash...so it can't be fetched by ordered values (blockNum)?

Comment (AskAlexSharov) [reply core/rawdb/accessors_chain.go:94]: Ah. Ok.

**Comment by sudeepdino008** [comment]
@somnathb1 added cache to store the heap; it'll return the sorted values when queried.

**Review by somnathb1 — CHANGES_REQUESTED** [review]
Please fix lint. Rest looks good.

**Review by somnathb1 — APPROVED** [review]

**Pull request merged**